sdks/fabric/interoperation-node-sdk/src/AssetManager.ts

```diff
@@ -0,0 +1,523 @@
+{
+    const lockInfoHTLC = new assetLocksPb.AssetLockHTLC();
+    lockInfoHTLC.setHashbase64(hashValue);
+    lockInfoHTLC.setExpirytimesecs(expiryTimeSecs);
+    lockInfoHTLC.setTimespec(assetLocksPb.AssetLockHTLC.TimeSpec.EPOCH)
+    const lockInfoHTLCSerialized = lockInfoHTLC.serializeBinary();
+    const lockInfo = new assetLocksPb.AssetLock();
+    lockInfo.setLockmechanism(assetLocksPb.LockMechanism.HTLC);
+    lockInfo.setLockinfo(lockInfoHTLCSerialized);
+    return lockInfo.serializeBinary().toString();
+}
+
+// Create an asset claim structure
+function createAssetClaimInfoSerialized(hashPreimage)
+{
+    const claimInfoHTLC = new assetLocksPb.AssetClaimHTLC();
+    claimInfoHTLC.setHashpreimagebase64(Buffer.from(hashPreimage).toString('base64'));
+    const claimInfoHTLCSerialized = claimInfoHTLC.serializeBinary();
+    const claimInfo = new assetLocksPb.AssetClaim();
+    claimInfo.setLockmechanism(assetLocksPb.LockMechanism.HTLC);
+    claimInfo.setClaiminfo(claimInfoHTLCSerialized);
+    return claimInfo.serializeBinary().toString();
+}
+
+// Create a SHA-256 hash over an ASCII string
+function createSHA256HashBase64(preimage)
+{
+    return crypto.createHash('sha256').update(preimage).digest('base64');
+}
+
+
+/**
+ * First/second step of a Hashed Time Lock Contract
+ * - Lock a unique asset instance using a hash
+ **/
+const createHTLC = async (
+    contract: Contract,
+    assetType: string,
+    assetID: string,
+    recipientECert: string,
+    hashValue: string,
+    expiryTimeSecs: number,
+): Promise<{ preimage: any; result: any }> => {
+
+    if (!contract)
+    {
+        logger.error("Contract handle not supplied");
+        return { preimage: "", result: false };
+    }
+    if (!assetType)
+    {
+        logger.error("Asset type not supplied");
+        return { preimage: "", result: false };
+    }
+    if (!assetID)
+    {
+        logger.error("Asset ID not supplied");
+        return { preimage: "", result: false };
+    }
+    if (!recipientECert)
+    {
+        logger.error("Recipient ECert not supplied");
+        return { preimage: "", result: false };
+    }
+
+    let defaultExpiryTimeSecs = 0, preimage = "";
+    if (hashValue && hashValue.length > 0)
+    {
+        defaultExpiryTimeSecs = 5 * 60     // 5 mins: this is the 't' used to time out the second contract of the HTLC pair
+    }
+    else
+    {
+        defaultExpiryTimeSecs = 10 * 60     // 10 mins: this is the '2t' used to time out the first contract of the HTLC pair
+
+        // Create a random preimage of 20 bytes
+        for (let i = 0 ; i < 20 ; i++)
+        {
+            preimage += String.fromCharCode(Math.floor(Math.random() * 256))
+        }
+
+        // Hash the preimage
+        hashValue = createSHA256HashBase64(preimage);
+    }
+    const currTimeSecs = Math.floor(Date.now()/1000);   // Convert epoch milliseconds to seconds
+    if (expiryTimeSecs <= currTimeSecs)
```

Comment:
I don't think time in the context of HTLCs is an implementation detail that the user can be oblivious to. While it would be ideal from a UX perspective if the user did not have to care about such parameters, this specific parameter dictates the behaviour of the user and is one they have to be conscious of (i.e. they have to ensure to manually perform certain actions within the bounds of this parameter or risk forfeiting the exchange). For this reason, I think the value of this parameter should be a deliberate choice by the user rather than an arbitrary default that they might not know about. In future, as we automate part of the currently manual steps of the HTLC flow, we can consider making these details transparent to the user

Reply:
While I agree that time is an even more sensitive parameter in protocols that have long confirmation times and probabilistic finality, it is still an important one even in permissioned networks, where we might take shorter confirmation times and instant finality for granted. The reason is that a good portion of what that time specifies is the pace at which a person is expected to action each step of the flow. This is something the party would want to be explicit about. Otherwise, we risk making undue assumptions i.e. that the party is trigger ready to execute each step.

An example of where such an assumption might not hold for instance is when considering the party is a large enterprise, with defined internal approval workflows that involve multiple parties signing off before any action that commits the organization is performed. In a solution, we recently developed, for instance, every transaction that comes to the blockchain required an off-chain approval workflow that adheres to the Four-Eye or Six-Eye principle (i.e. 2 parties or 3 parties in the organization sign-off respectively). If an HTLC agreement involving significant amounts of money is set up, it could be that each step of the HTLC process involves multiple off-chain approvals, before being submitted on-chain. This basically means a significant portion of the HTLC flow operates on slower human time, potentially involving a number of parties. A parameter that dictates the pace of this reality has to be deliberately set and agreed upon.

The other minor thing I would add is that from an engineering perspective if one provides a parameter to a function (albeit an invalid one) and the function silently ignores it and sets its own, it would seem to me that the behavior of the function violates the principle of least surprise :-)

Reply:
Done.